File: dart_lang/lib/a_solid/b_open_close/b_post_refactor/payment_processor.dart
import 'package:dart_lang/a_solid/b_open_close/b_post_refactor/payment_strategy/payment_strategy.dart';
import 'package:dart_lang/a_solid/helper_classes/helper_classes.dart';

class PaymentProcessor {
  final PaymentStrategy _paymentStrategy;
  PaymentProcessor(PaymentStrategy paymentStrategy)
      : _paymentStrategy = paymentStrategy;

  void processPayment(Order order, Payment payment) {
    print("Processing payment of order ${order.name}");
    print("Issuing payment for amount ${order.price}");
    _paymentStrategy.processPayment(order.price);
  }
}

File: Python/SOLID/open_close/solution/PaymentProcessor.py
from PaymentStrategy import *
from Python.SOLID.single_res.solution.order import Order

class PaymentProcessor:
    def __init__(self, payStrategy: PaymentStrategy):
        self.payStrategy = payStrategy

    def process_payment(self, order: Order):
        print(f"Processing payment for order: {order.get_name()}")
        print(f"Issuing payment for amount: {order.get_price()}")

        self.payStrategy.processPayment(order.get_price())

# Create products
if __name__ == "__main__":
    # Create products (orders)
    electronics_product = Order("Electronics", 100)
    clothing_product = Order("Clothing", 50)
    books_product = Order("Books", 30)

    # Choose the strategy
    mastercard_strategy = MasterCardPaymentStrategy()
    americanexpress_strategy = AmericanExpressPaymentStrategy()
    paypal_strategy = PaypalPaymentStrategy()

    # Create a payment processor and process payments
    processor = PaymentProcessor(mastercard_strategy)
    processor.process_payment(electronics_product)

    processor = PaymentProcessor(americanexpress_strategy)
    processor.process_payment(clothing_product)

    processor = PaymentProcessor(paypal_strategy)
    processor.process_payment(books_product)